source/extensions/clusters/aggregate/cluster.cc

```diff
@@ -0,0 +1,148 @@
+  //    C_0 [P_0, P_1, P_2]
+  //    C_1 [P_0, P_1]
+  //    C_2 [P_0, P_1, P_2, P_3]
+  // The linearization result is:
+  //    [C_0.P_0, C_0.P_1, C_0.P_2, C_1.P_0, C_1.P_1, C_2.P_0, C_2.P_1, C_2.P_2, C_2.P_3]
+  // and the traffic will be distributed among these priorities.
+  for (const auto& cluster : clusters_) {
+    if (deleting(cluster)) {
+      continue;
+    }
+    auto tlc = cluster_manager_.get(cluster);
+    if (tlc == nullptr) {
+      continue;
+    }
+
+    int priority_in_current_cluster = 0;
+    for (const auto& host_set : tlc->prioritySet().hostSetsPerPriority()) {
+      if (!host_set->hosts().empty()) {
+        priority_set.updateHosts(
+            next_priority_after_linearizing++, Upstream::HostSetImpl::updateHostsParams(*host_set),
+            host_set->localityWeights(), host_set->hosts(), {}, host_set->overprovisioningFactor());
+        priority_to_cluster.emplace_back(std::make_pair(priority_in_current_cluster, tlc));
+      }
+      priority_in_current_cluster++;
+    }
+  }
+
+  return std::make_pair(std::move(priority_set), std::move(priority_to_cluster));
+}
+
+void Cluster::startPreInit() {
+  for (const auto& cluster : clusters_) {
+    auto tlc = cluster_manager_.get(cluster);
+    if (tlc == nullptr) {
+      continue;
+    }
+
+    // Add callback for clusters initialized before aggregate cluster.
+    tlc->prioritySet().addMemberUpdateCb(
+        [this, cluster](const Upstream::HostVector&, const Upstream::HostVector&) {
+          ENVOY_LOG(info, "member update for cluster '{}'", cluster);
+          refresh();
+        });
+  }
+  refresh();
+  handle_ = cluster_manager_.addThreadLocalClusterUpdateCallbacks(*this);
+
+  onPreInitComplete();
+}
+
+void Cluster::refresh(std::function<bool(const std::string&)>&& deleting) {
+  auto pair = linearizePrioritySet(std::move(deleting));
+  if (pair.first.hostSetsPerPriority().empty()) {
+    if (load_balancer_ != nullptr) {
+      load_balancer_.reset();
+    }
+  } else {
+    load_balancer_ = std::make_shared<LoadBalancerImpl>(*this, std::move(pair));
+  }
+}
+
+void Cluster::onClusterAddOrUpdate(Upstream::ThreadLocalCluster& cluster) {
+  if (std::find(clusters_.begin(), clusters_.end(), cluster.info()->name()) != clusters_.end()) {
+    ENVOY_LOG(info, "update or add cluster '{}'", cluster.info()->name());
+    refresh();
+    cluster.prioritySet().addMemberUpdateCb(
+        [this](const Upstream::HostVector&, const Upstream::HostVector&) { refresh(); });
+  }
+}
+
+void Cluster::onClusterRemoval(const std::string& cluster_name) {
+  //  The onClusterRemoval callback is called before the thread local cluster was removed. There
+  //  will be a dangling pointer to the thread local cluster if delete cluster is not skipped.
+  if (std::find(clusters_.begin(), clusters_.end(), cluster_name) != clusters_.end()) {
+    ENVOY_LOG(info, "remove cluster '{}'", cluster_name);
```

Comment:
id include some indication that this is talking about a cluster being removed from the aggregate cluster